### ISSUE TITLE
Dispatch doesn't seem to complete until listener finishes?

### ISSUE THREAD

**Issue by bryannaegele** [goatslacker/alt#350]
TL;DR
Calling a transition on react router in the store listener of a view throws a "Cannot dispatch in the middle of a dispatch" error when the route you are transitioning to has a view which invokes a source fetch on another store in componentDidMount.

I'm seeing a problem when a component mounts (with a store fetch call in componentDidMount) raising a "cannot dispatch in the middle of a dispatch" error.

Basic setup is my header component is listening to a user store. Logging in calls a user login action which dispatches success/failure. The header's onUserChange handler I have for the store then looks at the incoming state. If the user is now authenticated, it calls a transition on react router to the user's home page. The page being transitioned to has a component that does a fetch on another store in componentDidMount.  This throws an error about being in the middle of a dispatch.

```javascript
export default class HeaderMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      user: UserStore.getState().user
    };
  }
  static getProps() {
    return {};
  }
  componentDidMount = () => {
    UserStore.listen(this.onUserChange);
  }

  componentWillUnmount = () => {
    UserStore.unlisten(this.onUserChange);
  }

  onUserChange = (nextState) => {
    let user = nextState.user;
    let currentState = this.state.user;

    // not currently authenticated, becomes authenticated
    if(currentState.get('authenticated') === false && user.get('authenticated')) {
      this.setState({
        user: user
      });
      // invoking these transitions here or in the callback of the setState does not make a difference
      this.context.router.transitionTo('myProjects', { organization: user.get('realm')});
    } else if (currentState.get('authenticated') && user.get('authenticated') === false) {
      // currently authenticated, becomes unauthenticated
      this.setState({
        user: user
      });
      this.context.router.transitionTo('login');
    } else if( !Immutable.is(currentState, user)) {
      this.setState({
        user: user
      });
    }
  }

  render() {
    var {router} = this.context;
    let user = this.state.user;
    let classes = cn('headerMenu', styles.this);
    if (user.get('authenticated')) {
      return <div className={classes}>
        <LoggedInMenu user={user} />
      </div>;
    } else {
      return <div className={classes}>
        <LoggedOutMenu />
      </div>;
    }
  }
}

HeaderMenu.contextTypes = {
  router: React.PropTypes.func.isRequired
};
```
And the page transitioning to...

```javascript
export default class MyProjectsPage extends React.Component {
  constructor(props) {
    super(props);
    var projects = ProjectListStore.getState().projects;
    this.state = {
      projectsLoading: ProjectListStore.isLoading(),
      projects: projects
    };
  }
  static getProps() {
    return {};
  }
  componentDidMount = () => {
    console.log('mounting project page');
    // error throws here
    ProjectListStore.fetchProjects(this.props.params.organization);
    ProjectListStore.listen(this.onProjectsChange);
  }
  componentWillUnmount = () => {
    ProjectListStore.unlisten(this.onProjectsChange);
  }
  onProjectsChange = (nextState) => {
    if (!nextState.isLoading) {
      this.setState({
        projects: nextState.projects
      });
    }
  }
  render() {
    var classes = cn('myProjectsPage', 'std-page-container', styles.this);
    return (
      <section className={classes}>
        <header className='std-page-container-header'>
          <h1 className='std-page-container-title'>My Projects</h1>
          <div className='std-button'>
            + Add Project
          </div>
        </header>
        <ProjectGrid projects={this.state.projects} />
      </section>
    );
  }
}
```

User Actions
```javascript
manualLogin(credentials) {
    let { username, password, realm } = credentials;
    UserWebAPIUtils.login(username, password, realm)
      .then((response) => {
        this.actions.loginSuccess(response);
      })
      .catch((err) => {
        this.actions.loginFailure(err);
      });
  }

  loginSuccess(data) {
    this.dispatch({
      username: data.user.username,
      token: data.id,
      tokenTTL: data.ttl,
      tokenCreated: data.created,
      realm: data.user.realm,
      id: data.user.id,
      email: data.user.email
    });
  }

  loginFailure(err) {
    this.dispatch(err);
  }
```
![image](https://cloud.githubusercontent.com/assets/5731285/8347273/da95ebd6-1acb-11e5-96a4-14b638824bae.png)

Any ideas?

A related issue was the only way I found this was catching the error in login failure of the login method of the util. It would appear the dispatch in loginSuccess throws an error that then bubbled up to the UserWebAPIUtils's promise, thus invoking the catch of it and triggering the login failure. Any tips there would be greatly appreciated.

**Comment by goatslacker** [comment]
Try this out:

```js
const alt = new Alt({
  batchingFunction: React.addons.batchedUpdate
});
```

**Comment by bryannaegele** [comment]
Same error.

app.js
```javascript
import Alt from 'alt';
import React from 'react/addons';
const isomorphicRouterRenderer = require('./utils/IsomorphicRouterRenderer');
const routes = require('./routes.jsx');
const alt = new Alt({
  batchingFunction: React.addons.batchedUpdate
});

module.exports = isomorphicRouterRenderer(alt, routes);
```

isomorphic renderer from webpack react-starter
```javascript
import Iso from 'iso';
import React from 'react/addons';
import Router from 'react-router';
import routes from 'routes';

function IsomorphicRouterRenderer(alt) {
  if (typeof window === 'undefined') {
    return function (state, url) {
      var markup;
      Router.run(routes, url, function (Handler) {
        alt.bootstrap(state);
        var content = React.renderToString(React.createElement(Handler));
        markup = Iso.render(content, alt.takeSnapshot());
        alt.flush();
      });
      return markup;
    };
  } else {
    Iso.bootstrap(function (state, _, container) {
      alt.bootstrap(state);
      Router.run(routes, Router.HistoryLocation, function (Handler) {
        var node = React.createElement(Handler);
        React.render(node, container);
      });
    });
  }
}

module.exports = IsomorphicRouterRenderer;
```

**Comment by goatslacker** [comment]
More context on the React.addons.batchedUpdates thing: https://github.com/BinaryMuse/fluxxor/commit/51f7651cfae6cfe78216c347787f46b4e5237979

One thing you can do which will definitely fix it is to use `defer()` http://alt.js.org/docs/actions/#actiondefer

**Comment by bryannaegele** [comment]
Where would you put the defer? The piece causing the issue is when calling a method exposed on the store from a source. Would I just want to expose that through a normal action for now as a workaround?

Thanks for the link for context, btw. I'm coming from Reflux, so basic Flux principles are still new for me. Love this library and Iso, btw. The Reflux'ish patterns have made it relatively painless save this.

**Comment by goatslacker** [comment]
I hastily mentioned defer which wouldn't be of any use to you...you could wrap that call in a setTimeout or setImmediate but that's gross...

Do you have a loading action setup in the source? That gets called synchronously. If you remove it, problems go away. Is there a reason you're handling loading?

**Comment by bryannaegele** [comment]
That is a fantastic question. I had it because I thought it was required when I glanced at the initial code block on the docs. I thought it was required boilerplate to set the isLoading state and expose the isLoading() method on the store. Looking at the source I see that isn't and is just a convenience point.

I removed it and it works great now. 

I'll make a PR to make the doc more idiot proof for the more casual reader such as myself on required/optional methods. I also added what may be helpful info on ways to always fetch remotely.

**Comment by ghost** [comment]
@goatslacker I do have a loading in my source, for informing the store (and thereby components) that something is getting fetched and I can show a spinner. 

But now I run into the exact same problem with `Cannot dispatch in the middle of a dispatch`.

What's the best practice here?
